shrinklvm.py

```diff
@@ -0,0 +1,1012 @@
+    assert(label_magic == 'LABELONE')
+    assert(sector_xl == label_sector)
+    assert(offset_xl == 32)
+    assert(lvm_type == 'LVM2 001')
+    assert(calc_crc(label_data[20:]) == crc_xl) # CRC is calculated from offset_xl to the end of the sector
+
+    # Read PV header
+    parse_offset = offset_xl
+    lvm.pv_id, lvm.dev_size = struct.unpack_from('<32sQ', label_data, offset=parse_offset)
+    parse_offset += 40
+
+    # Read data areas
+    while True:
+        offset, size = struct.unpack_from('<QQ', label_data, offset=parse_offset)
+        parse_offset += 16
+        if offset == 0:
+            break
+
+        lvm.da.append(Area(offset, size))
+        logger.log('Found LVM data area: {}, {}'.format(offset, size))
+
+    # Read metadata areas
+    while True:
+        offset, size = struct.unpack_from('<QQ', label_data, offset=parse_offset)
+        parse_offset += 16
+        if offset == 0:
+            break
+        lvm.mda.append(Area(offset, size))
+        logger.log('Found LVM metadata area: {}, {}'.format(offset, size))
+
+    # Read PV header extension
+    lvm.pv_ext_version, lvm.pv_ext_flags = struct.unpack_from('<II', label_data, offset=parse_offset)
+    parse_offset += 8
+
+    # Read bootloader areas
+    while True:
+        offset, size = struct.unpack_from('<QQ', label_data, offset=parse_offset)
+        parse_offset += 16
+        if offset == 0:
+            break
+        lvm.ba.append(Area(offset, size))
+        logger.log('Found LVM bootloader area: {}, {}'.format(offset, size))
+
+    # Read metadata area header
+    assert(len(lvm.mda) > 0)
+    lvm.file.seek(lvm.mda[0].offset, os.SEEK_SET)
+    buf = lvm.file.read(SECTOR_SIZE)
+
+    checksum_xl, magic, version, start, size = struct.unpack_from('<I16sIQQ', buf)
+    assert(calc_crc(buf[4:]) == checksum_xl)
+    assert(magic == ' LVM2 x[5A%r0N*>')
+    assert(version == 1)
+    assert(start == lvm.mda[0].offset)
+    assert(size == lvm.mda[0].size)
+
+    # Read raw locations of metadata
+    parse_offset = 40
+    while True:
+        offset, size, checksum, flags = struct.unpack_from('<QQII', buf, offset=parse_offset)
+        parse_offset += 24
+        if offset == 0:
+            break
+        lvm.raw_locs.append(Area(offset, size, checksum, flags))
+        logger.log('Found LVM config: {}, {}'.format(offset, size))
+
+    assert(len(lvm.raw_locs) > 0)
+
+    # Now read the actual metadata
+    # The metadata is stored in a circular buffer.
+    wrap = max(0, lvm.raw_locs[0].offset + lvm.raw_locs[0].size - lvm.mda[0].size)
+
+    lvm.file.seek(lvm.raw_locs[0].offset + start, os.SEEK_SET)
+    buf = lvm.file.read(lvm.raw_locs[0].size - wrap)
+
+    if wrap:
+        lvm.file.seek(start + MDA_HEADER_SIZE, os.SEEK_SET)
+        buf += lvm.file.read(wrap)
+
+    assert(calc_crc(buf) == lvm.raw_locs[0].checksum)
+
+    lvm.text_config = buf[:-1] # ignore NUL terminator
```

Comment:
Is it mandatory to have the `NUL` terminator? Does the standard specify some encoding for this text?

Reply:
I haven't seen documentation that describes the details of the on-disk format. I believe the implementation always uses a NUL terminator: https://github.com/lvmteam/lvm2/blob/bb384f84880a183ee58a18935851eece25cdada4/lib/format_text/export.c#L1053

The implementation also seems to use C strings without an explicit encoding. Using Python strings is probably the most suitable data type to use.

Reply:
I will change this to strip any amount of NULs from the end in case it is not present.